### PR TITLE
start integrating TokenVault

### PR DESCRIPTION
**Pull request by YouStillAlive**

### CODE REVIEW

File: contracts/BaseProvider/BaseLockDealProvider.sol [hunk 1]
```diff
@@ -17,8 +21,8 @@ contract BaseLockDealProvider is BaseLockDealModifiers, IProvider {
         address token,
         uint256[] memory params
     ) public returns (uint256 poolId) {
-        poolId = lockDealNFT.mint(owner, token);
-        _registerPool(poolId, params);
+        _registerPool(poolId, token, params);
+        poolId = lockDealNFT.mint(owner, token, msg.sender, params[0]);
```
Comment (Lomet) [inline contracts/BaseProvider/BaseLockDealProvider.sol:25]: we need to get poolId first, then make the _registerPool with the right poolId

File: contracts/DealProvider/DealProvider.sol [hunk 2]
```diff
@@ -19,13 +20,11 @@ contract DealProvider is DealProviderModifiers, IProvider {
...
         notZeroAddress(token)
-        notZeroAmount(params[0])
         validParamsLength(params.length, currentParamsTargetLenght)
         returns (uint256 poolId)
     {
-        poolId = lockDealNFT.mint(owner, token);
-        poolIdToDeal[poolId].leftAmount = params[0];
-        poolIdToDeal[poolId].token = token;
+        _registerPool(poolId, token, params);
+        poolId = lockDealNFT.mint(owner, token, msg.sender, params[0]);
```
Comment (Lomet) [inline contracts/DealProvider/DealProvider.sol:27]: need swap order

File: contracts/interface/IVaultFactory.sol [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+interface IVaultFactory{
+    function CreateNewVault(address tokenAddress) external returns(address);
```
Comment (Lomet) [inline contracts/interface/IVaultFactory.sol:5]: I don't think we needed this here

**Review by Lomet — APPROVED** [review]
Looks good, few small change to merge it

File: contracts/LockDealNFT/LockDealState.sol [hunk 4]
```diff
@@ -4,10 +4,14 @@ pragma solidity ^0.8.0;
...
+import "poolz-helper-v2/contracts/interfaces/IVaultManager.sol";
 
 abstract contract LockDealState is ERC721Enumerable {
-    Counters.Counter public _tokenIdCounter;
+    Counters.Counter public tokenIdCounter;
+    IVaultManager public vaultManager;
 
     mapping(uint256 => address) public poolIdToProvider;
+    mapping(uint256 => uint256) public poolIdToVaultId;
+    mapping(address => uint256) public tokenToVaultId;
```
Comment (Lomet) [inline contracts/LockDealNFT/LockDealState.sol:15]: I think we don't need this

**Review by Lomet — COMMENTED** [review]
Looks good !
I think we can delete some code now :)

File: contracts/BaseProvider/BaseLockDealProvider.sol [hunk 5]
```diff
@@ -23,7 +27,7 @@ contract BaseLockDealProvider is
         address token,
         uint256[] memory params
     ) public returns (uint256 poolId) {
-        poolId = lockDealNFT.mint(owner, token, params[0]);
+        poolId = lockDealNFT.mint(owner,  token, msg.sender, params[0]);
```
Comment (Lomet) [inline contracts/BaseProvider/BaseLockDealProvider.sol:30]: I think its better to explicitly pass the address,
so the handle can open a pool for other provider

File: contracts/LockDealNFT/LockDealNFTModifiers.sol [hunk 6]
```diff
@@ -36,11 +31,41 @@ abstract contract LockDealNFTModifiers is LockDealState, Ownable {
...
+    modifier approvedAmount(address token, address from, uint256 amount) {
+        _approvedAmount(token, from, amount);
+        _;
+    }
+
     function _notZeroAddress(address _address) private pure {
         require(_address != address(0x0), "Zero Address is not allowed");
     }
 
+    function _onlyContract(address contractAddress) private view {
```
Comment (Lomet) [inline contracts/LockDealNFT/LockDealNFTModifiers.sol:43]: why do we need this ?

Comment (YouStillAlive) [reply contracts/LockDealNFT/LockDealNFTModifiers.sol:43]: providers are only contracts

File: test/TimedLockDealProvider.js [hunk 7]
```diff
@@ -18,7 +18,10 @@ describe("Timed Lock Deal Provider", function (accounts) {
...
         const TimedLockDealProvider = await ethers.getContractFactory("TimedLockDealProvider")
         const ERC20Token = await ethers.getContractFactory("ERC20Token")
-        lockDealNFT = await LockDealNFT.deploy()
+        const MockVaultManager = await ethers.getContractFactory("MockVaultManager")
+        const mockVaultManagger = await MockVaultManager.deploy()
+        await mockVaultManagger.deployed()
+        lockDealNFT = await LockDealNFT.deploy(mockVaultManagger.address)
         await lockDealNFT.deployed()
         dealProvider = await DealProvider.deploy(lockDealNFT.address)
         await dealProvider.deployed()
```
Comment (Lomet) [inline test/TimedLockDealProvider.js:27]: we can extract the base logic to Test_helper file, and just use it in each file

File: contracts/BaseProvider/BaseLockDealProvider.sol (hunk 5 above)
Comment (YouStillAlive) [reply contracts/BaseProvider/BaseLockDealProvider.sol:30]: Hmm, I don't quite get it. 
Pass the provider's address to the mint function?

File: contracts/LockDealNFT/LockDealNFTModifiers.sol (hunk 6 above)
Comment (YouStillAlive) [reply contracts/LockDealNFT/LockDealNFTModifiers.sol:43]: when we use internal functions inside modifiers less gas is used

**Pull request merged**